### PR TITLE
DEV-4213: pin deps, unpin dev deps

### PR DESCRIPTION
**Pull request by leo-dcfa**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

N/A

#### What's this PR do?

- Pin regular dependencies
- Unpin dev dependencies
- Fix one file that was failing on pre-commit
- Fix test failures after deps updates - Stripe tests were generating dates with microseconds whilst datetime.datetime.fromtimestamp wasn't - leading to test failures, since the two values would be different. Setting the microseconds to 0 did the trick. Additionally, pydantic_core is no longer imported from pydantic, but as a standalone import.

#### Why are we doing this? How does it help us?

This will make it so we can do a regular poetry update; pre-commit autoupdate once per sprint to keep deps up-to-date

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

N/A

#### Have automated unit tests been added? If not, why?

N/A

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

N/A

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-4213]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

N/A

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

N/A

[DEV-4213]: https://news-revenue-hub.atlassian.net/browse/DEV-4213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/tests/test_serializers.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from django.utils import timezone
 
 import pydantic
+import pydantic_core
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_serializers.py:10]: **Question**

This change seems out of scope. What's reasoning?

File: apps/contributions/tests/test_stripe_contributions_provider.py [hunk 2]
```diff
@@ -233,7 +233,7 @@ def test_card(self, payment_method, expected, pi_for_valid_one_time_factory, moc
     def test_last_payment_date_when_no_invoice(self, pi_without_invoice):
         assert StripePaymentIntent(pi_without_invoice).last_payment_date == datetime.datetime.fromtimestamp(
             int(pi_without_invoice.created), tz=datetime.timezone.utc
-        )
+        ).replace(microsecond=0)
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_stripe_contributions_provider.py:236]: Unclear why this change and similar ones below required.

**Review by benjaminEwhite — COMMENTED** [review]
Have a couple questions of small questions (the answer to which I'm assuming is that changes were required to make pre-commit happy, but want to understand).

File: apps/contributions/tests/test_serializers.py (hunk 1 above)
Comment (leo-dcfa) [reply apps/contributions/tests/test_serializers.py:10]: I was also surprised with this one - when I ran `poetry update` the test that references `pydantic.pydantic_core.Url` started failing. I noticed that pydantic_core got bumped from `2.10.1`to `2.14.3`, check the documentation and imported directly instead.

File: apps/contributions/tests/test_stripe_contributions_provider.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/tests/test_stripe_contributions_provider.py:236]: **blocking request**

worth adding a code comment about why this is required.


**non-blocking question**

do you know why the test started failing now?

**Review by benjaminEwhite — APPROVED** [review]

File: apps/contributions/tests/test_stripe_contributions_provider.py (hunk 2 above)
Comment (leo-dcfa) [reply apps/contributions/tests/test_stripe_contributions_provider.py:238]: Comment added. 

Concerning why, I don't know, some dependency update led to it

File: pyproject.toml [hunk 3]
```diff
@@ -90,7 +90,7 @@ bandit = "1.7.5"
 black = "23.9.1"
 colorama = "*"
 coverage = "*"
-dateparser = "^1.1.8"
+dateparser = "1.1.8"
```
Comment (x110dc) [inline pyproject.toml:93]: Why is this pinned and not set to `*`?

**Review by x110dc — CHANGES_REQUESTED** [review]
One dep doesn't look correct. 

Adding more detail to the "what does this PR do" section I think could've avoided some of the back-and-forth for the non-obvious changes here.

**Review by x110dc — APPROVED** [review]

**Pull request merged**